### PR TITLE
Add Node affinity

### PR DESCRIPTION
**Pull request by bdunne**
For OpenShift 4.14 support

Part of #988

### CODE REVIEW

**Comment by bdunne** [comment]
@nasark Please also review.  The refactoring was inspired by your work in #993.  Do you think the `miqutils` code should continue to live in `api/v1alpha1` or should create a `api/v1` for it?

**Comment by nasark** [comment]
> @nasark Please also review. The refactoring was inspired by your work in https://github.com/ManageIQ/manageiq-pods/pull/993. Do you think the miqutils code should continue to live in api/v1alpha1 or should create a api/v1 for it?

In my opinion since its technically just a refactor I would say to keep it in `v1alpha1`. Is there any benefit at the moment to creating a `v1` API? We would also need to support multiple versions in the CRD if thats the case

File: manageiq-operator/config/manager/manager.yaml
Comment (nasark) [inline manageiq-operator/config/manager/manager.yaml:23]: Do we also want `s390x` and `ppc64le` like in downstream?

Comment (nasark) [reply manageiq-operator/config/manager/manager.yaml:23]: In downstream we have a separate setNodeAffinity function that adds the other arches to the Kafka CR

**Comment by bdunne** [comment]
> > @nasark Please also review. The refactoring was inspired by your work in #993. Do you think the miqutils code should continue to live in api/v1alpha1 or should create a api/v1 for it?
> 
> In my opinion since its technically just a refactor I would say to keep it in `v1alpha1`. Is there any benefit at the moment to creating a `v1` API? We would also need to support multiple versions in the CRD if thats the case

- I was thinking of `v1` just for the `miqutils`, not the CRD and everything else yet. (That's much more involved)
- v1 implies stable library, but maybe this isn't there yet.

File: manageiq-operator/config/manager/manager.yaml
Comment (bdunne) [reply manageiq-operator/config/manager/manager.yaml:23]: > Do we also want s390x and ppc64le like in downstream?

No, we only build x86_64 upstream

>  I'm not sure how this will work in downstream though if we have additional arches.

We don't use this file downstream.  The downstream operator will deploy the manageiq operator with the correct values.

File: manageiq-operator/api/v1alpha1/miqutils/node_affinity.go
Comment (bdunne) [inline manageiq-operator/api/v1alpha1/miqutils/node_affinity.go:13]: I added a commit to default to `amd64`

**Review by nasark — APPROVED** [review]
LGTM, just a note that Strimzi Kafka will require a different method of setting node affinity (similar to downstream). But that can be handled in https://github.com/ManageIQ/manageiq-pods/pull/1005 I suppose

**Review by Fryguy — APPROVED** [review]
LGTM, also want @nasark approval before merge.

**Comment by Fryguy** [comment]
@bdunne How do we set the node affinities for the workloads? (workers and now also the automaton worker's jobs?)

**Comment by bdunne** [comment]
> @bdunne How do we set the node affinities for the workloads? (workers and now also the automaton worker's jobs?)
> 
> I also updated the OP to say this is part of #988, but wasn't sure if it "completed" #988 or if we need other PRs (such as the node affinities for the workloads)

This will discover the node affinities applied to the operator and apply the same affinities to httpd, postgres, memcached and the orchestrator.  I'll be opening a PR shortly that does the same tihng inside the orchestrator for the deployments that it creates.

**Comment by Fryguy** [comment]
@bdunne Can we merge this now or do we have to wait for the core orchestrator change?

Second question.  If we run on an older OpenShift, will these nodeAffinities cause issues, or are they just ignored on older OpenShifts

**Comment by bdunne** [comment]
This can be merged before the orchestrator changes.

My test cluster is 4.13.4 and it's working fine there. I can't find when `kubernetes.io/arch` label was introduced, but I see references to it over 3 years ago.

**Comment by Fryguy** [comment]
Skipping backport to `quinteros`, because it is already in the branch.

**Comment by Fryguy** [comment]
Skipping backport to `quinteros`, because it is already in the branch.

**Pull request merged**
